modules/sc-mesh-secure-deployment/src/2_0/features/cbma/macsec/macsec.py

```diff
@@ -0,0 +1,69 @@
+import subprocess
+import sys
+import os
+import random
+import threading
+
+script_dir = os.path.dirname(__file__) # path to macsec directory
+sys.path.insert(0, f'{script_dir}/../')
+from tools.custom_logger import CustomLogger
+
+logger_instance = CustomLogger("macsec")
+class Macsec:
+    def __init__(self, level, interface, macsec_encryption):
+        self.level = level # Macsec level: "lower" or "upper"
+        self.interface = interface
+        self.available_ports = set(range(1, 2**16)) # 1 to 2^16-1
+        self.used_ports = {} # client_mac: port
+        self.available_ports_lock = threading.Lock()
+        self.macsec_encryption = macsec_encryption  # Flag to set macsec encrypt on or off
+        self.logger = logger_instance.get_logger()
+
+    def set_macsec_tx(self, client_mac, my_macsec_key, my_port):
+        # Sets up macsec link and adds tx channel
+        macsec_interface = self.get_macsec_interface_name(client_mac)
+        try:
+            subprocess.run(["ip", "link", "add", "link", self.interface, macsec_interface, "type", "macsec", "port", str(my_port), "encrypt", self.macsec_encryption, "cipher", "gcm-aes-256"], check=True)
```

Comment:
Please add `send_sci off end_station on` to reduce the overhead from 32 to 24 :) This is possible because we create MACsec secure channels only between two MAC addresses ("end stations") as opposed to the more common use of it whereby the MACsec source might not be the same as the Ethernet source so this needs to be clarified with the Secure Channel Identifier.